core/parse.ml

```diff
@@ -8,28 +8,35 @@ open Lexing
    occurs.
 *)
 
-type 'a grammar = (Lexing.lexbuf -> Parser.token) -> Lexing.lexbuf -> 'a
-
 class source_code = SourceCode.source_code
 
-(* Read and parse Links source code from the source named `name' via
-   the function `infun'.
-*)
-let read : context:Lexer.lexer_context
-        -> ?nlhook:(unit -> unit)
-        -> parse:('intermediate grammar)
-        -> infun:(bytes -> int -> int)
-        -> name:string
-        -> 'result * source_code =
-fun ~context ?nlhook ~parse ~infun ~name ->
-  let code = new source_code in
-  let lexbuf = {(from_function (code#parse_into infun))
+module type LexerSig = sig
+  type token
+  type lexer_context
+  type 'a grammar = (Lexing.lexbuf -> token) -> Lexing.lexbuf -> 'a
+
+  val lexer : lexer_context
+           -> newline_hook:(unit -> unit)
+           -> (Lexing.lexbuf -> token)
+  val fresh_context : unit -> lexer_context
+end
+
+module Reader (Lex : LexerSig) = struct
```

Comment:
The functor name `Reader` is a little generic, though natural given that the main function was already called `read`.  Maybe something like `Tokenizer`/`tokenize`?  Or maybe I'm mistaking what is going on here,

Reply:
All the names where a bit ad-hoc. I've really ran out of ideas for names, especially in the presence of very similar names like `Parse` and `Parser`. Better ideas are certainly welcome.

Reply:
I suppose `Scanner` fits the bill.